src/helpers/data_channel.ts

```diff
@@ -0,0 +1,146 @@
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with Threema Web. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * A flow-controlled (sender side) data channel.
+ */
+export class FlowControlledDataChannel {
+    // Logging
+    private readonly logTag: string = '[FlowControlledDataChannel]';
+    private $log: ng.ILogService;
+
+    // Data channel
+    public readonly dc: RTCDataChannel;
+    public readonly highWaterMark: number;
+
+    // Flow control mechanism
+    private _ready: Future<void> = new Future();
+
+    /**
+     * Create a flow-controlled (sender side) data channel.
+     *
+     * @param $log The Angular logging service
+     * @param dc The data channel to be flow-controlled
+     * @param lowWaterMark The low water mark unpauses the data channel once
+     *   the buffered amount of bytes becomes less or equal to it.
+     * @param highWaterMark The high water mark pauses the data channel once
+     *   the buffered amount of bytes becomes greater or equal to it.
+     */
+    public constructor(
+        $log: ng.ILogService, dc: RTCDataChannel,
+        lowWaterMark: number = 262144, highWaterMark: number = 1048576,
+    ) {
+        this.$log = $log;
+        this.dc = dc;
+        this.highWaterMark = highWaterMark;
+
+        // Allow writing
+        this._ready.resolve();
+
+        // Unpause once low water mark has been reached
+        this.dc.bufferedAmountLowThreshold = lowWaterMark;
+        this.dc.onbufferedamountlow = () => {
+            if (!this._ready.done) {
+                this.$log.debug(this.logTag, `${this.dc.label} resumed (buffered=${this.dc.bufferedAmount})`);
+                this._ready.resolve();
+            }
+        };
+    }
+
+    /**
+     * A future whether the data channel is ready to be written on.
+     */
+    public get ready(): Future<void> {
+        return this._ready;
+    }
+
+    /**
+     * Write a message to the data channel's internal buffer for delivery to
+     * the remote side.
+     *
+     * Important: Before calling this, the `ready` Promise must be awaited.
+     *
+     * @param message The message to be sent.
+     * @throws Error in case the data channel is currently paused.
+     */
+    public write(message: Uint8Array): void {
+        // Throw if paused
+        if (!this._ready.done) {
+            throw new Error('Unable to write, data channel is paused!');
+        }
+
+        // Try sending
+        // Note: Technically we should be able to catch a TypeError in case the
+        //       underlying buffer is full. However, there are other reasons
+        //       that can result in a TypeError and no browser has implemented
+        //       this properly so far. Thus, we use a well-tested high water
+        //       mark instead and try to never fill the buffer completely.
+        this.dc.send(message);
```

Comment:
We don't check the size of `message`, so if someone would send a 50 MiB message the DC would still close, right?

Reply:
The data channel has to do that check. If it doesn't, it's a bug (like in Chrome). If we want to work around this, then yeah, we would have to check it. However, it's never going to happen since we do chunking.